client/state/posts/reducer.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import indexBy from 'lodash/collection/indexBy';
+
+/**
+ * Internal dependencies
+ */
+import {
+	POST_RECEIVE,
+	POSTS_RECEIVE,
+	POSTS_REQUEST,
+	POSTS_REQUEST_SUCCESS,
+	POSTS_REQUEST_FAILURE
+} from 'state/action-types';
+
+/**
+ * Tracks all known post objects, indexed by post global ID.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function items( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_RECEIVE:
+			state = Object.assign( {}, state, {
+				[ action.post.global_ID ]: action.post
+			} );
+			break;
+
+		case POSTS_RECEIVE:
+			state = Object.assign( {}, state, indexBy( action.posts, 'global_ID' ) );
+			break;
+	}
+
+	return state;
+}
+
+/**
+ * Returns the updated site posts state after an action has been dispatched.
+ * The state reflects a mapping of site ID, post ID pairing to global post ID.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function sitePosts( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_RECEIVE:
+		case POSTS_RECEIVE:
+			let posts = action.post ? [ action.post ] : action.posts;
+			posts.forEach( ( post ) => {
+				state = Object.assign( {}, state );
+				if ( ! state[ post.site_ID ] ) {
+					state[ post.site_ID ] = {};
+				}
+
+				state[ post.site_ID ][ post.ID ] = post.global_ID;
+			} );
+			break;
+	}
+
+	return state;
+}
+
+/**
+ * Returns the updated post query state after an action has been dispatched.
+ * The state reflects a mapping of site ID to active queries.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function siteQueries( state = {}, action ) {
+	switch ( action.type ) {
+		case POSTS_REQUEST:
+		case POSTS_REQUEST_SUCCESS:
+		case POSTS_REQUEST_FAILURE:
+			const { siteId, posts } = action;
+			const query = JSON.stringify( action.query );
+
+			state = Object.assign( {}, state );
+			if ( ! state[ siteId ] ) {
+				state[ siteId ] = {};
+			}
+
+			state[ siteId ][ query ] = {
+				fetching: POSTS_REQUEST === action.type
+			};
+
+			if ( posts ) {
+				state[ siteId ][ query ].posts = posts.map( ( post ) => {
+					return post.global_ID;
+				} );
```

Comment:
>this looks like a good place to reduce the extra syntax and clear up the code IMO

I went back-and-forth on this a few times, and it came down to the difference of a few extra characters in the line-length and a worry about whether it might hurt legibility. I do think the shorthand is nicely concise, so I'll see about changing it.